apis/bigquery/v1beta1/dataset_reference.go

```diff
@@ -0,0 +1,187 @@
+	// Should be in the format "projects/<projectID>/locations/<location>/datasets/<datasetID>".
+	External string `json:"external,omitempty"`
+
+	// The name of a BigQueryDataset resource.
+	Name string `json:"name,omitempty"`
+
+	// The namespace of a BigQueryDataset resource.
+	Namespace string `json:"namespace,omitempty"`
+
+	parent *BigQueryDatasetParent
+}
+
+// NormalizedExternal provision the "External" value for other resource that depends on BigQueryDataset.
+// If the "External" is given in the other resource's spec.BigQueryDatasetRef, the given value will be used.
+// Otherwise, the "Name" and "Namespace" will be used to query the actual BigQueryDataset object from the cluster.
+func (r *BigQueryDatasetRef) NormalizedExternal(ctx context.Context, reader client.Reader, otherNamespace string) (string, error) {
+	if r.External != "" && r.Name != "" {
+		return "", fmt.Errorf("cannot specify both name and external on %s reference", BigQueryDatasetGVK.Kind)
+	}
+	// From given External
+	if r.External != "" {
+		if _, _, err := ParseBigQueryDatasetExternal(r.External); err != nil {
+			return "", err
+		}
+		return r.External, nil
+	}
+
+	// From the Config Connector object
+	if r.Namespace == "" {
+		r.Namespace = otherNamespace
+	}
+	key := types.NamespacedName{Name: r.Name, Namespace: r.Namespace}
+	u := &unstructured.Unstructured{}
+	u.SetGroupVersionKind(BigQueryDatasetGVK)
+	if err := reader.Get(ctx, key, u); err != nil {
+		if apierrors.IsNotFound(err) {
+			return "", k8s.NewReferenceNotFoundError(u.GroupVersionKind(), key)
+		}
+		return "", fmt.Errorf("reading referenced %s %s: %w", BigQueryDatasetGVK, key, err)
+	}
+	// Get external from status.externalRef. This is the most trustworthy place.
+	actualExternalRef, _, err := unstructured.NestedString(u.Object, "status", "externalRef")
+	if err != nil {
+		return "", fmt.Errorf("reading status.externalRef: %w", err)
+	}
+	if actualExternalRef == "" {
+		return "", k8s.NewReferenceNotReadyError(u.GroupVersionKind(), key)
+	}
+	r.External = actualExternalRef
+	return r.External, nil
+}
+
+// New builds a BigQueryDatasetRef from the Config Connector BigQueryDataset object.
+func NewBigQueryDatasetRef(ctx context.Context, reader client.Reader, obj *BigQueryDataset) (*BigQueryDatasetRef, error) {
+	id := &BigQueryDatasetRef{}
+
+	// Get Parent
+	projectRef, err := refsv1beta1.ResolveProject(ctx, reader, obj, obj.Spec.ProjectRef)
+	if err != nil {
+		return nil, err
+	}
+	projectID := projectRef.ProjectID
+	if projectID == "" {
+		return nil, fmt.Errorf("cannot resolve project")
+	}
+	location := obj.Spec.Location
+	id.parent = &BigQueryDatasetParent{ProjectID: projectID, Location: valueOf(location)}
+
+	// Get desired ID
+	resourceID := valueOf(obj.Spec.ResourceID)
+	if resourceID == "" {
+		resourceID = obj.GetName()
+	}
+	if resourceID == "" {
+		return nil, fmt.Errorf("cannot resolve resource ID")
+	}
+
+	// Use approved External
+	externalRef := ""
```

Comment:
good catch. This bypasses the immutable validation, @xiaoweim could you elaborate more on this?

Reply:
There is no externalRef field in the "status" of this resource. I see that by having the externalRef as an empty string, it can bypass the immutable field `project`, `location` and `resourceID` validation. 
Is it necessary to add this field to status or it is okay if we can do this validation through other field such as `selfLink`?

Reply:
We prefer using `status.externalRef` because the base direct reconciler builds additional logic around it.  For `selfLink`, we want to keep it (I know it's a dup), could you add a comment saying that we recommend users to use the `status.externalRef` instead?

Reply:
I see, sure I can add the recommendation. Let me add `externalRef` here in the status and update the PR. Thank you for the explanation!